controllers/k8ssandra/replication.go

```diff
@@ -50,24 +74,209 @@ func (r *K8ssandraClusterReconciler) checkSystemReplication(ctx context.Context,
 // updateReplicationOfSystemKeyspaces ensures that the replication for the system_auth,
 // system_traces, and system_distributed keyspaces is up to date. It ensures that there are
 // replicas for each DC and that there is a max of 3 replicas per DC.
-func (r *K8ssandraClusterReconciler) updateReplicationOfSystemKeyspaces(ctx context.Context, kc *api.K8ssandraCluster, dc *cassdcapi.CassandraDatacenter, remoteClient client.Client, logger logr.Logger) result.ReconcileResult {
-	managementApiFacade, err := r.ManagementApi.NewManagementApiFacade(ctx, dc, remoteClient, logger)
-	if err != nil {
-		logger.Error(err, "Failed to create ManagementApiFacade")
-		return result.Error(err)
+func (r *K8ssandraClusterReconciler) updateReplicationOfSystemKeyspaces(
+	ctx context.Context,
+	kc *api.K8ssandraCluster,
+	mgmtApi cassandra.ManagementApiFacade,
+	logger logr.Logger) result.ReconcileResult {
+
+	if recResult := r.versionCheck(ctx, kc); recResult.Completed() {
+		return recResult
 	}
 
 	keyspaces := []string{"system_traces", "system_distributed", "system_auth"}
-	replication := cassandra.ComputeReplication(3, kc.Spec.Cassandra.Datacenters...)
+	datacenters := cassandra.GetDatacentersForSystemReplication(kc)
+	replication := cassandra.ComputeReplication(3, datacenters...)
 
 	logger.Info("Preparing to update replication for system keyspaces", "replication", replication)
 
 	for _, ks := range keyspaces {
-		if err := managementApiFacade.EnsureKeyspaceReplication(ks, replication); err != nil {
+		if err := mgmtApi.EnsureKeyspaceReplication(ks, replication); err != nil {
 			logger.Error(err, "Failed to update replication", "keyspace", ks)
 			return result.Error(err)
 		}
 	}
 
 	return result.Continue()
 }
+
+// updateUserKeyspacesReplication updates the replication factor of user-defined keyspaces.
+// The K8ssandraCluster must specify the k8ssandra.io/dc-replication in order for any
+// updates to be applied. The annotation can specify multiple DCs but only the DC just
+// added will be considered for replication changes. For example, if dc2 is added to a
+// cluster that has dc1 and if the annotation specifies changes for both dc1 and dc2, only
+// changes for dc2 will be applied. Replication for all user-defined keyspaces must be
+// specified; otherwise an error is returned. This is required to avoid surprises for the
+// user.
+func (r *K8ssandraClusterReconciler) updateUserKeyspacesReplication(
+	ctx context.Context,
+	kc *api.K8ssandraCluster,
+	dc *cassdcapi.CassandraDatacenter,
+	mgmtApi cassandra.ManagementApiFacade,
+	logger logr.Logger) result.ReconcileResult {
+
+	jsonReplication := annotations.GetAnnotation(kc, api.DcReplicationAnnotation)
+	if jsonReplication == "" {
+		logger.Info(api.DcReplicationAnnotation + " not set. Replication for user keyspaces will not be updated")
```

Comment:
Should this be logger.Error() or even a EventRecorder.Eventf() ? I don't think users will notice this kind of notice that everything is fine (result.Continue()) but we didn't actually do anything they requested.

Reply:
Or if this is normal behavior, set the logging to debug or just not log anything?